src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java

```diff
@@ -52,4 +57,28 @@ public String authenticateUser(String username, String password) {
         return BEARER_AUTH_TYPE + " " + tokenExchangeResponse.getAccessToken();
     }
 
+    public TokenExchangeResponse exchangeCode(ExchangeCodeRequest exchangeCodeRequest) {
+        return idamApi.exchangeCode(exchangeCodeRequest);
+    }
+
+    public GeneratePinResponse generatePin(GeneratePinRequest pinRequest, String authorization) {
+        return idamApi.generatePin(pinRequest, authorization);
+    }
+
+    public AuthenticateUserResponse authenticatePinUser(String pin, String clientId, String redirectUrl, String state) {
+        AuthenticateUserResponse pinUserCode;
+        Response response =  idamApi.authenticatePinUser(pin, clientId, redirectUrl, state);
+        String code = getCodeFromRedirect(response);
+        pinUserCode = new AuthenticateUserResponse(code);
+
+        return pinUserCode;
+    }
+
+    private String getCodeFromRedirect(Response response) {
```

Comment:
Will this be handled as a 400 bad request? The main issue I see with this client is it results in a 500 error on failures which doesn't help service teams identify problems

Reply:
this bit of the code is for tests only I think, which will likely have full request response logging enabled (and should probably enable it if it's not already enabled)

Reply:
This isn't for tests, it is to extract the code from a redirect URL and return it, as we do in many projects (CMC, Divorce). It is part of the OAuth flow

Reply:
This method fits into existing use cases nicely @nikola-naydenov-hmcts

Reply:
I can wrap in a try/catch to avoid the 500

Reply:
I think when you return a response it doesn't throw an exception and you have full control, but you'll want to test it out

Reply:
Ok sure, but I am going to assume the missing error decoder in this code means that service teams will override and create their own, as the number of messages we get with, "hey this request didn't work 500 error is idam broken?"... Anyway I'll resolve this conversation